modules/navier_stokes/include/userobjects/RhieChowMassFlux.h

```diff
@@ -0,0 +1,226 @@
+   * pressure equation plus Rhie-Chow interpolation.
+   */
+  void computeHbyA(bool verbose);
+
+  bool hasFaceSide(const FaceInfo & fi, const bool fi_elem_side) const override;
+
+protected:
+  void setupCellVolumes();
+
+  /// Populate the face values of the H/A and 1/A fields
+  void
+  populateCouplingFunctors(const std::vector<std::unique_ptr<NumericVector<Number>>> & raw_hbya,
+                           const std::vector<std::unique_ptr<NumericVector<Number>>> & raw_Ainv);
+
+  /**
+   * Check the block consistency between the passed in \p var and us
+   */
+  template <typename VarType>
+  void checkBlocks(const VarType & var) const;
+
+  /// The \p MooseMesh that this user object operates on
+  MooseMesh & _moose_mesh;
+
+  /// The \p libMesh mesh that this object acts on
+  const libMesh::MeshBase & _mesh;
+
+  /// The dimension of the mesh, e.g. 3 for hexes and tets, 2 for quads and tris
+  const unsigned int _dim;
+
+  /// The thread 0 copy of the pressure variable
+  MooseLinearVariableFVReal * const _p;
+
+  /// The thread 0 copy of the x-velocity variable
+  std::vector<MooseLinearVariableFVReal *> _vel;
+
+  LinearFVAnisotropicDiffusion * _p_diffusion_kernel;
+
+  /**
+   * A map functor from faces to $HbyA_{ij} = (A_{offdiag}*\mathrm{(predicted~velocity)} -
+   * \mathrm{Source})_{ij}/A_{ij}$. So this contains the off-diagonal part of the system matrix
+   * multiplied by the predicted velocity minus the source terms from the right hand side of the
+   * linearized momentum predictor step.
+   */
+  FaceCenteredMapFunctor<Real, std::unordered_map<dof_id_type, Real>> _HbyA_flux;
+
+  /**
+   * We hold on to the cell-based HbyA vectors so that we can easily reconstruct the
+   * cell velocities as well.
+   */
+  std::vector<std::unique_ptr<NumericVector<Number>>> _HbyA_raw;
+
+  /**
+   * A map functor from faces to $(1/A)_f$. Where $A_i$ is the diagonal of the system matrix
+   * for the momentum equation.
+   */
+  FaceCenteredMapFunctor<RealVectorValue, std::unordered_map<dof_id_type, RealVectorValue>> _Ainv;
+
+  /**
+   * We hold on to the cell-based 1/A vectors so that we can easily reconstruct the
+   * cell velocities as well.
+   */
+  std::vector<std::unique_ptr<NumericVector<Number>>> _Ainv_raw;
+
+  /**
+   * A map functor from faces to mass fluxes which are used in the advection terms.
+   */
+  FaceCenteredMapFunctor<Real, std::unordered_map<dof_id_type, Real>> _face_mass_flux;
+
+  /**
+   * Functor describing the density of the fluid
+   */
+  const Moose::Functor<Real> & _rho;
+
+  /// Pointers to the nonlinear system(s) corresponding to the momentum equation(s)
+  std::vector<LinearSystem *> _momentum_systems;
+
+  /// Numbers of the momentum system(s)
+  std::vector<unsigned int> _momentum_system_numbers;
+
+  /// Global numbers of the momentum system(s)
+  std::vector<unsigned int> _global_momentum_system_numbers;
+
+  /// Pointers to the momentum equation implicit system(s)
```

Comment:
same systems, different pointer type only right? might be worth saying

Reply:
actually I see the solution is copied twice, once in each?

Reply:
`LinearSystem` (in MOOSE) has a `LinearImplicitSystem` (libmesh) member. The solutions are stored in the libmesh system.

Reply:
I extended the comments on the member variables